### DISCUSSION TITLE
0-9 Scanlist behavior

### DISCUSSION THREAD

**Discussion by 121tech** [AubsUK/uv-k5-firmware-custom#21] [Q&A] [answered]
Chose your firmware fork specifically for the multiple scanlist.
I'm pretty sure when I installed v0.02 that if I placed memory channels in a scanlist bank, enabled multiple scanlist banks, then set the radio to scan then all the active scanlist banks got swept.

ISSUE: 
Just updated to v0.06 and, maybe I'm remembering wrong as I appear to have lost the ability to scan multiple scanlist banks. The radio will now only scan the lowest enabled scanlist number and refuses to scan any other enabled scanlist bank. The screen displays the stars and dashes but the scan ignores the star enabled scanlist. 

I tried reflashing back to v0.02 but it has the same new behavior. Reflashed back on v0.06, tried to perform a factory reset (hold PTT + SIDE BUTTON 1 while turning on the radio, after 'RELEASE ALL KEYS' navigate to Menu 77 Reset ), now 68.68 under v.006, but the scan function remains only able to scan the lowest enabled scanbank.

**Comment by AubsUK** [comment]
That sounds pretty unusual.  My understanding and some guessing would be:
* You factory reset using menu 68 and selected ALL.
* You've got all stars in 0-9 showing that all ScanLists are enabled, but no number showing for the active ScanList.

That most likely means the currently selected bank doesn't have any channels stored.

If you go to a scan memory channel, and select Menu 14, it'll show which ScanList(s) the channel is in.  I'd check a couple of channels to make sure they are populated.

Have you tried loading the memories via CHIRP?

**Reply by AubsUK** [reply]
Incidentally, if it's only 'scanning' the FIRST programmed channel (generally M1), that means no channels have been added to any of the active ScanLists.

e.g. My ScanList 0 doesn't have any channels in it.  if I enable only ScanList 0, it takes me to the first channel only.  I could work on displaying "none", but that might be overkill, but if it's the reason, and gives a clear answer, it might be worth it.

**Comment by 121tech** [comment]
Using CHIRP to set up channels, and "VIEW: Show extra Fields" to set the scanlists, then sending the setup to the radio.
The photo image of the uv-k5(8) shows what I see, scanlists 0, 1, 2 selected (0**-------) but the screen only displays the channels in scanlist 0 being scanned. If I press zero, then scanlist 1 becomes active but only scanlist 1 is scanned and scanlist 2 continues to be ignored.  

I believe the radio used to show 0**-------, then * 1 * -------,` then **2-------, and all channels in those scanlist banks were actively swept and that showed on the display. For whatever reason that is no longer happening. currently it only scans the lowest active scanlist digit and ignores the other active scanlist entries.  

![image](https://github.com/AubsUK/uv-k5-firmware-custom/assets/29210691/1ba96abd-bd2c-48d9-a440-d2187e52af59)
![image](https://github.com/AubsUK/uv-k5-firmware-custom/assets/29210691/4852458b-aac8-4599-be1a-f6c6eb92fbbb)

**Reply by AubsUK** [reply]
Perfect, thanks for the update.

From the photo, it looks like it is scanning the channels in ScanList 0 (the blurry channel number indicates that to me as it's switching channel).

I think if you disable ScanList 0, it'll start scanning all channels in ScanList 1.  If you re-enable ScanList 0, it'll continue scanning all channels in ScanList 1, i.e. it won't go back to ScanList 0.

If you have time and feel competent enough (I don't want you to do something you're not comfortable with), could you have a look at this comment https://github.com/AubsUK/uv-k5-firmware-custom/discussions/9#discussioncomment-8881009. It would be useful if you could launch CHIRP with my module loaded and download from radio, then check if and confirm if you see 1s or 0s in an unused memory location?

From your screenshot, Memory 8 is empty, that would correspond to Memory Location 7

**Comment by 121tech** [comment]
In Developer Mode, Chirp restarted, fresh Download from radio:

![image](https://github.com/AubsUK/uv-k5-firmware-custom/assets/29210691/f57b23b1-469c-4e08-af3f-d9636e6bf9dd)

![image](https://github.com/AubsUK/uv-k5-firmware-custom/assets/29210691/69aa87b5-d24e-4a05-bd69-9eed668b223c)

Channel 10 in the above config shows as empty, Browser channelscanlist "9" shows Scanlist0: '1'  <-- not a zero

**Comment by 121tech** [comment]
I'm guessing this was caused when I used Chirp to move some channels around and writing that back to the radio after installing v0.06.  The channels didn't get cleared out and the previous data got written to the radio.

I think my best move here would be to reset the radio, load from the radio, copy the channels from the previous config to where I want them in the new config, then write that back to the radio. This should remove any hidden left-over channel config that's causing my scanlist glitch.

**Comment by 121tech** [comment]
FYI, blank channels 10 11 12 (channelscanlists 9 10 11) had HEX '1' in scanlist0, blank channel 15 (channelscanlists 14) had HEX '1' in scanlist1

**Reply by AubsUK** [reply]
I read your previous reply, looked at the screenshot and thought it all looked good, I was replying!  Perhaps I'm getting tired.  As you pointed out here, yes indeed, Memory Channel 10 is empty, but it's memory location (9) is not cleared properly.

I would expect CHIRP to write back empty memory locations if any were changed, but I've got something to look into at least, perhaps it isn't doing as I'd expect.

I'll also look into why it's using a memory location that isn't configured.  Perhaps I can fix that sooner.

If you're happy to re-do your settings, you can factory reset again (hold `PTT` + `SIDE BUTTON 1` while turning on the radio, after 'RELEASE ALL KEYS' navigate to Menu 68 `Reset` then select `ALL`), then you can read from radio and copy/paste all the channels from the older IMG to the blank IMG (as detailed in this comment https://github.com/AubsUK/uv-k5-firmware-custom/discussions/9#discussioncomment-8887560) and upload to radio.

Thanks for finding this, I'll move it over to be an issue instead of a discussion, as I thought I'd fixed the checking if it is a configured channel, but perhaps not.

**Reply by AubsUK** [reply]
I suspect if you saved a memory to Channel 10 on the radio itself, it would already be in ScanList 1.

**Reply by 121tech** [reply]
Yep. I did a Reset ALL on the radio, had CHIRP read in the fresh reset blank radio, copied over my channels from the previous config to the new one, set the channels scanlist so it only used '0' for the Air channels and '1' for the EMS channels, then wrote that back to the radio ...and the scan function is back to working as you wrote it, moving from active scanlist 0 to 1 and back to 0

So, dirty data in CHIRP Memories after editing channels with scanlist entries back to 'blank' (but not really left blank), then the dirty data written to the radio causing the scanlist function to work incorrectly. Ouch.

**Reply by AubsUK** [reply]
Very ouch! Sorry it's caused you issues, but thank you for finding it and your perseverance to get it working.  I'll work on the firmware to ignore any non-channels (I thought I had already, but clearly only fixed one part of it) and also the CHIRP module to write back cleared memory locations.  I've created [Issue 22](https://github.com/AubsUK/uv-k5-firmware-custom/issues/22) for it.

**Comment by AubsUK** [comment, marked as answer]
I've fixed this issue in [v00.07](https://github.com/AubsUK/uv-k5-firmware-custom/releases/tag/v00.07).  Thank you for your help.

I've also fixed it in the CHIRP module, but I've not updated that online yet.  Even though it's not really needed as the radio will know not to use the deleted channel, I still want it to be right.

**Reply by AubsUK** [reply]
I think I've also fixed this in the latest CHIRP driver too, just download it from https://github.com/AubsUK/uvk5-chirp-driver/blob/main/uvk5_Aubs.py
